src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs

```diff
@@ -0,0 +1,496 @@
+        public async Task EndTextWriterValueScopeAsync()
+        {
+            if (!IsWritingJson)
+            {
+                this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+                await this.textWriter.DisposeAsync();
+                await this.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+
+                CheckIfSeparatorNeeded();
+                CheckIfManualValueAtArrayStart();
+            }
+        }
+
+        /// <summary>
+        /// Whether the current TextWriter is writing JSON
+        /// </summary>
+        /// <returns></returns>
+        private bool IsWritingJson
+        {
+            get
+            {
+                return String.IsNullOrEmpty(this.currentTextWriterContentType) || this.currentTextWriterContentType.StartsWith(MimeConstants.MimeApplicationJson, StringComparison.Ordinal);
+            }
+        }
+
+        /// <summary>
+        /// Represents a TextWriter implementation for writing UTF-8 JSON data in an OData context.
+        /// </summary>
+        internal class ODataUtf8JsonTextWriter : TextWriter
+        {
+            ODataUtf8JsonWriter jsonWriter = null;
+            private char[] buffer;
+            int numOfCharsNotWrittenFromPreviousChunk = 0;
+
+            public ODataUtf8JsonTextWriter(ODataUtf8JsonWriter jsonWriter)
+            {
+                this.jsonWriter = jsonWriter;
+            }
+
+            public override Encoding Encoding => throw new NotImplementedException();
+
+            /// <summary>
+            /// Flushes any buffered data to the underlying stream synchronously.
+            /// </summary>
+            public override void Flush()
+            {
+                this.jsonWriter.Flush();
+            }
+
+            /// <summary>
+            /// Flushes any buffered data to the underlying stream asynchronously.
+            /// </summary>
+            /// <param name="cancellationToken">A cancellation token to observe while waiting for the flush operation to complete.</param>
+            /// <returns>A task representing the asynchronous flush operation.</returns>
+            public override async Task FlushAsync()
+            {
+                await this.jsonWriter.FlushAsync();
+            }
+
+            /// <summary>
+            /// Disposes the object.
+            /// </summary>
+            /// <param name="disposing">True if called from Dispose; false if called form the finalizer.</param>
+            protected override void Dispose(bool disposing)
+            {
+                this.Flush();
+
+                // If there are unprocessed chars, encode and write them as the final block.
+                ReadOnlySpan<char> charsNotProcessedFromPreviousChunk = this.buffer.AsSpan().Slice(0, numOfCharsNotWrittenFromPreviousChunk);
+
+                if (!charsNotProcessedFromPreviousChunk.IsEmpty)
+                {
+                    if (!this.jsonWriter.IsWritingJson)

```

Comment:
I think it would be valuable to add a comment explaining why we don't perform escaping when the content type is not JSON. Like explicitly mention that we assume the input to be already valid JSON with all escaping handled by the caller. Maybe that should be specified in the `StartTextWriterScope` documentation.

Reply:
we don't perform escaping when the content type is JSON not the other way round. I'll add a comment though.